### ISSUE TITLE
Sub-command support

### ISSUE THREAD

**Issue by ocavue** [alexflint/go-arg#75]
*This issue is a **feature discussion***

I really enjoy this project. The API is extremely simple and the help message is detailed. In fact, I think go-arg is the most elegant choice in [this list](https://github.com/avelino/awesome-go#command-line).

However, go-arg doesn't support sub-command, which is a common feature used by many command line tools (much common than nested flags). If go-arg supports it, I think go-args could cover 80%~90% use cases of a command line tool.

> Many third-party argument parsing libraries are geared for writing sophisticated command line interfaces. The excellent `codegangsta/cli` is perfect for working with multiple sub-commands and nested flags, but is probably overkill for a simple script with a handful of flags.

Thanks again for such an awesome project!

**Comment by alexflint** [comment]
Hey @ocavue - thank you for the very kind words :)

Seems reasonable! I've thought about this before and would be open to implementing it. I was thinking of an approach using a separate struct for each subcommand, something like:
```
type Cmd1 struct {
  Abc string
}
type Cmd2 struct {
   Xyz string `arg:"required"`
}
type Args struct {
  Cmd1 *Cmd1 arg:"subcommand:cmd1"
  Cmd2 *Cmd2 arg:"subcommand:cmd2"
  GlobalFlag string
}
```
This would allow any of the following:
```
./myprogram cmd1
./myprogram cmd1 --abc=value
./myprogram --globalflag=aaa cmd1
./myprogram --globalflag=aaa cmd1 --abc=value
./myprogram --globalflag=aaa  // fine to omit subcommand... to enforce that a subcommand must be specified just check manually
```
but not any of these
```
./myprogram cmd1 cmd2  // max one subcommand
./myprogram cmd2  // because --xyz is required
```

Also:
- Ban mixing of `positional` with `subcommand` because it would get too complicated to assign positionals between top-level and nested commands. Can lift this in the future if we find a robust solution.
- need to decide how help text will be formatted (e.g. does `./myprogram cmd1 --help` print help for all subcommands or just cmd1? How about `./myprogram --help cmd1`?

**Comment by ocavue** [comment]
> Ban mixing of `positional` with `subcommand`

Agree.


> How about `./myprogram --help cmd1`?

I did a little research to some command line tools that I use. There are 3 ways to print help message for a sub-command:

```
myprogram cmd1 --help
myprogram help cmd1
myprogram --help cmd1
```

- `git` `npm` support all these 3 ways.
- `go` `pip` support first 2 ways.
- [`cloc`](https://github.com/AlDanial/cloc), [`jupyter`](https://jupyter.org/) only support the first way.
- `apt-get` supports none of these ways …...

So, I think we can only support `./myprogram cmd1 --help` since it's the most common way to do it and we want to keep this script as simple as possible.



> does `./myprogram cmd1 --help` print help for all subcommands or just cmd1?

All command line tools I listed above (except `apt-get`) print help message for cmd1, so I think just cmd1 will be fine.

**Comment by ocavue** [comment]
BTW, speaking of "print help for all subcommands or just cmd1", [docopt](<https://github.com/docopt/docopt.go>) has a very different behavior. It well print all help for root-command and for all sub-commands:

```
$ naval_fate --help

Usage:
  naval_fate ship new <name>...
  naval_fate ship <name> move <x> <y> [--speed=<kn>]
  naval_fate ship shoot <x> <y>
  naval_fate mine (set|remove) <x> <y> [--moored|--drifting]
  naval_fate -h | --help
  naval_fate --version

Options:
  -h --help     Show this screen.
  --version     Show version.
  --speed=<kn>  Speed in knots [default: 10].
  --moored      Moored (anchored) mine.
  --drifting    Drifting mine


$ naval_fate ship --help

Usage:
  naval_fate ship new <name>...
  naval_fate ship <name> move <x> <y> [--speed=<kn>]
  naval_fate ship shoot <x> <y>
  naval_fate mine (set|remove) <x> <y> [--moored|--drifting]
  naval_fate -h | --help
  naval_fate --version

Options:
  -h --help     Show this screen.
  --version     Show version.
  --speed=<kn>  Speed in knots [default: 10].
  --moored      Moored (anchored) mine.
  --drifting    Drifting mine

```


For a small script, I found it has great readability. But it lacks description for sub-command itself and the help message is too long for command-line tools with a lot of sub-commands, like `git` or `docker`.

**Comment by alexflint** [comment]
Implemented in #82
